ballista/rust/scheduler/src/state/task_manager.rs

```diff
@@ -376,11 +375,78 @@ impl<T: 'static + AsLogicalPlan, U: 'static + AsExecutionPlan> TaskManager<T, U>
         .await
     }
 
+    pub(crate) async fn cancel_job(
+        &self,
+        job_id: &str,
+        executor_manager: &ExecutorManager,
+    ) -> Result<()> {
+        let lock = self.state.lock(Keyspace::ActiveJobs, "").await?;
+
+        let running_tasks = self
+            .get_execution_graph(job_id)
+            .await
+            .map(|graph| graph.running_tasks())
+            .unwrap_or_else(|_| vec![]);
+
+        info!(
+            "Cancelling {} running tasks for job {}",
+            running_tasks.len(),
+            job_id
+        );
+
+        self.fail_job_inner(lock, job_id, "Cancelled".to_owned())
+            .await?;
+
+        let mut tasks: HashMap<&str, Vec<protobuf::PartitionId>> = Default::default();
+
+        for (partition, executor_id) in &running_tasks {
+            if let Some(parts) = tasks.get_mut(executor_id.as_str()) {
+                parts.push(protobuf::PartitionId {
+                    job_id: job_id.to_owned(),
+                    stage_id: partition.stage_id as u32,
+                    partition_id: partition.partition_id as u32,
+                })
+            } else {
+                tasks.insert(
+                    executor_id.as_str(),
+                    vec![protobuf::PartitionId {
+                        job_id: job_id.to_owned(),
+                        stage_id: partition.stage_id as u32,
+                        partition_id: partition.partition_id as u32,
+                    }],
+                );
+            }
+        }
+
+        for (executor_id, partitions) in tasks {
+            if let Ok(mut client) = executor_manager.get_client(executor_id).await {
+                client
+                    .cancel_tasks(CancelTasksParams {
+                        partition_id: partitions,
+                    })
+                    .await?;
+            } else {
+                error!("Failed to get client for executor ID {}", executor_id)
+            }
+        }
+
+        Ok(())
+    }
+
```

Comment:
The performance might be a problem here. Multiple cancel task requests are issued sequentially.  Seems launch_task has the same problem. We can spawn tokio tasks to cancel/launch tasks concurrently.

Reply:
See below, the task cancellation itself is fast (only setting an `AtomicBoolean`). We're ultimately limited by the number of CPU cores available so I would worry that spawning thousands of tokio tasks to try and do the cancellations concurrently could actually cause a performance issue rather than alleviate one. If we did do it that way we would need a way to control the level of concurrency. In general, I think it's probably best to only try and optimize this once we've observed a performance issue related to it as I think the solution may be more complicated.

Reply:
We will not spawn thousands of tokio tasks here in most cases, the concurrency level would  depends on how many executor instances we have in the system. I think sending the cancel request to multiple executors concurrently is better than sending the request sequentially. The same for assign/launch tasks to executors.  Usually in a data center, a remote RPC call could take 5ms to 20 ms.

Reply:
Ultimately, might we want a `CancelTask(TaskIds[])` message so we can just send one message to each executor? Personally I'm not worried either way here since coroutines are just file handles - it's not like we're blocking a thread or spawning a thread.

Reply:
@avantgardnerio This is what is happening already. We build a `HashMap<ExecutorId, Vec<Task>>` and then call each executor only once with all tasks that need to be cancelled on that executor.